### ISSUE TITLE
MBASIC shuffles lines randomly

### ISSUE THREAD

**Issue by robinhood2014** [DJSures/NABU-LIB#4]
When I load and run one of my ports of Del Horn's "Golden Flutes and Great Escapes" text-based adventure games, it will sometimes crash with a bogus error message reading "Undefined line number." Upon LISTing the program, I find that MBASIC has shuffled a line in the program; that is, it has placed a line where it does not belong. For example, in "Treasure Hunt" (the annotation is my own):
```
...
950 IF B(CP)=10 THEN 1520
1690 IF C=6 THEN FR=FR-DC        'ERROR!
960 IF B(CP)=20 THEN 1560
...
```
Or again, in "The Golden Flute" (once again my own annotation):
```
...
5440 IF D=1 THEN PRINT"BROMBIRAN";:BR=0
5780 R=RND*9+1:IF R>8 THEN RETURN        'ERROR!
5450 IF D=2 THEN PRINT"DAGGLETTE";:DG=0
...
```
This also happens in "The Great Escape" and "Mars," but those instances aren't listed here.

I originally thought this to be an internal problem within MBASIC itself, so I wasn't sure if the problem could be fixed, as, although MBASIC is classified as "abandonware," it's still copyrighted by Microsoft and, as such, can't be modified. However, when I ran the same programs (with a minor modification in each) in GW-BASIC (a later version of MBASIC for DOS), I get no line shuffling whatsoever. This leads me to believe that there's no problem in MBASIC or GW-BASIC.

As the problem happens only intermittently, I'm usually able to work around it by reloading the offending program. Sometimes, however, the problem remanifests itself several times in a row.

I was intrigued by your DUMPRAM program (in A:1). It showed me that MBASIC doesn't do a very good job of cleaning up the NABU's memory when a NEW or CLEAR statement is executed.

I'm enclosing my C: drive (in a ZIP file) containing the game programs. Also included on the C: drive are text files containing the instructions and "HELP" subroutine from "Mars", high score files for "The Golden Flute", "The Great Escape" and "Mars", and a small program that clears these high scores. (There's also a menu program which I based on the Apple ]['s version of the compilation, but you can use it or lose it as you see fit.)
[C.zip](https://github.com/DJSures/NABU-LIB/files/11012462/C.zip)

**Comment by DJSures** [comment]
I updated the MBASIC on A:2 to 5.29. See if this fixes your issue. From what I read, there's a number of changes with 5.29. I'll have to look for a manual online to see the changes

**Comment by robinhood2014** [comment]
I'm sorry, no. This particular issue wasn't fixed with that released; I'm still getting instances of line shuffling.

**Comment by DJSures** [comment]
Hmm i'm not sure why those mbasic's do that. Have you tried any other basic versions? I don't have much experience with basic so i'm not sure if there's a line count max, or memory max or something. is there a way to show how much ram is free in basic?

**Comment by robinhood2014** [comment]
Yes, the command to do it is `PRINT FRE(0)`.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by robinhood2014** [comment]
I have ported all the GFGE games to BBC BASIC. They're in user area 9 of my C drive. I have noticed that BBC BASIC is significantly faster than MBASIC in loading and executing my programs, that my programs are significantly smaller in file size, and that there's no line shuffling at all!

However...

Sometimes each program will crash with a bogus error such as `No such line` or `No such variable`. To make sure I'm not going crazy, I'll `LIST` the offending line to make sure everything is in good working order; then, depending on the situation, I'll run `GOTO XXX` (where XXX is the line where the program crashed), and the program will keep going — or else I'll just do `RUN` again.

I just thought it was worth mentioning, to make sure the whole NABU community has a more enjoyable Cloud CP/M experience.
